File: app/app/src/main/java/com/shelter/shade/engine/Buildings.kt
package com.shelter.shade.engine

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const val DEFAULT_LEVEL_HEIGHT_M = 3.0   // metres per floor
const val DEFAULT_BUILDING_HEIGHT_M = 9.0 // fallback when no height/levels tag

// ---------------------------------------------------------------------------
// Data model
// ---------------------------------------------------------------------------

/**
 * Building footprint (lat/lon polygon) and height.
 *
 * [ring] elements are doubleArrayOf(lat, lon) — GeoJSON [lon,lat] is swapped
 * on load so that callers always work in (lat, lon) order.
 * [heightM]          ground-to-roof height in metres.
 * [heightEstimated]  true when height was inferred from levels or default.
 */
data class Building(
    val ring: List<DoubleArray>,
    val heightM: Double,
    val heightEstimated: Boolean = false,
    val osmId: String? = null,
) {
    /**
     * Returns [minLat, minLon, maxLat, maxLon].
     */
    fun bbox(): DoubleArray {
        var minLat = Double.MAX_VALUE
        var minLon = Double.MAX_VALUE
        var maxLat = -Double.MAX_VALUE
        var maxLon = -Double.MAX_VALUE
        for (pt in ring) {
            if (pt[0] < minLat) minLat = pt[0]
            if (pt[1] < minLon) minLon = pt[1]
            if (pt[0] > maxLat) maxLat = pt[0]
            if (pt[1] > maxLon) maxLon = pt[1]
        }
        return doubleArrayOf(minLat, minLon, maxLat, maxLon)
    }
}

// ---------------------------------------------------------------------------
// Height estimation
// ---------------------------------------------------------------------------

/**
 * Parse an OSM `height` tag value that may look like "12", "12 m", "12.5m",
 * "12 meters", "12 metres". Returns the float value in metres, or null if the
 * string cannot be parsed or is non-positive.
 */
private fun parseMeters(value: String): Double? {
    val s = value
        .trim()
        .lowercase()
        .replace("meters", "")
        .replace("metres", "")
        .replace("m", "")
        .trim()
    return try {
        val v = s.toDouble()
        if (v > 0) v else null
    } catch (_: NumberFormatException) {
        null
    }
}

/**
 * Estimate building height from OSM tags.
 *
 * Priority: `height` tag → `building:levels` / `levels` × [DEFAULT_LEVEL_HEIGHT_M]
 *           → [DEFAULT_BUILDING_HEIGHT_M].
 *
 * Returns Pair(heightM, estimated) where *estimated* is false only when a
 * concrete `height` tag was successfully parsed.
 */
fun estimateHeightM(tags: Map<String, String>): Pair<Double, Boolean> {
    val rawHeight = tags["height"]
    if (rawHeight != null) {
        val parsed = parseMeters(rawHeight)
        if (parsed != null) return Pair(parsed, false)
    }

    val levels = tags["building:levels"] ?: tags["levels"]
    if (levels != null) {
        val n = try { levels.trim().toDouble() } catch (_: NumberFormatException) { null }
        if (n != null && n > 0) {
            return Pair(n * DEFAULT_LEVEL_HEIGHT_M, true)
        }
    }

    return Pair(DEFAULT_BUILDING_HEIGHT_M, true)
}

// ---------------------------------------------------------------------------
// GeoJSON loader
// ---------------------------------------------------------------------------

private val _json = Json { ignoreUnknownKeys = true }

/**
 * Parse a GeoJSON FeatureCollection [text] into a list of [Building]s.
 *
 * Handles Polygon and MultiPolygon geometries. Only the exterior ring (first
 * ring) of each polygon is used. GeoJSON coordinates are [lon, lat]; they are
 * swapped to (lat, lon) as required by [Building.ring].
 *
 * Rings with fewer than 3 distinct points are skipped.
 */
fun loadGeoJson(text: String): List<Building> {
    val root = _json.parseToJsonElement(text).jsonObject
    val features = root["features"]?.jsonArray ?: return emptyList()
    return featuresToBuildings(features)
}

private fun featuresToBuildings(features: JsonArray): List<Building> {
    val buildings = mutableListOf<Building>()

    for (featureElem in features) {
        val feature = featureElem as? JsonObject ?: continue
        val geometry = feature["geometry"] as? JsonObject ?: continue
        val gtype = (geometry["type"] as? JsonPrimitive)?.content ?: continue
        val coords = geometry["coordinates"] as? JsonArray ?: continue

        val props = feature["properties"] as? JsonObject
        val tags: Map<String, String> = props
            ?.entries
            ?.mapNotNull { (k, v) ->
                val prim = v as? JsonPrimitive ?: return@mapNotNull null
                k to prim.content
            }
            ?.toMap()
            ?: emptyMap()

        val (height, estimated) = estimateHeightM(tags)

        // OSM id: prefer properties["id"], then feature["id"]
        val osmId: String? = run {
            val fromProps = (props?.get("id") as? JsonPrimitive)?.content
            val fromFeature = (feature["id"] as? JsonPrimitive)?.content
            fromProps ?: fromFeature
        }

        // Collect exterior rings (first ring of each polygon)
        val exteriorRings: List<JsonArray> = when (gtype) {
            "Polygon" -> {
                val firstRing = coords.getOrNull(0) as? JsonArray
                if (firstRing != null) listOf(firstRing) else emptyList()
            }
            "MultiPolygon" -> {
                coords.mapNotNull { polyElem ->
                    val poly = polyElem as? JsonArray ?: return@mapNotNull null
                    poly.getOrNull(0) as? JsonArray
                }
            }
            else -> continue
        }

        for (ringCoords in exteriorRings) {
            val ring = ringCoords.mapNotNull { ptElem ->
                val pt = ptElem as? JsonArray ?: return@mapNotNull null
                if (pt.size < 2) return@mapNotNull null
                val lon = (pt[0] as? JsonPrimitive)?.content?.toDoubleOrNull() ?: return@mapNotNull null
                val lat = (pt[1] as? JsonPrimitive)?.content?.toDoubleOrNull() ?: return@mapNotNull null
                doubleArrayOf(lat, lon)   // swap: GeoJSON is [lon,lat]
            }
            if (ring.size >= 3) {
                buildings.add(
                    Building(
                        ring = ring,
                        heightM = height,
                        heightEstimated = estimated,
                        osmId = osmId,
                    )
                )
            }
        }
    }

    return buildings
}

File: shade-engine/shade_engine/buildings.py
"""건물 모델 및 높이 추정.

그림자를 드리우는 캐스터의 1차 대상은 건물이다. OSM `building` 폴리곤의
`height`(미터) 태그를 우선 사용하고, 없으면 `building:levels` × 층고로 추정,
둘 다 없으면 기본값을 적용한다. 높이가 추정값인지 여부를 보존해 신뢰도 계산에 쓴다.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

DEFAULT_LEVEL_HEIGHT_M = 3.0  # 층당 기본 층고
DEFAULT_BUILDING_HEIGHT_M = 9.0  # 높이/층수 정보가 전혀 없을 때 가정(약 3층)


@dataclass(frozen=True)
class Building:
    """건물 footprint(위경도 폴리곤)와 높이.

    ring: [(lat, lon), ...] 외곽 폴리곤. 닫힘 여부는 무관(자동 처리).
    height_m: 지면 기준 높이(미터).
    height_estimated: 높이가 태그 직접값이 아니라 추정값이면 True → 신뢰도 하락 요인.
    """

    ring: tuple[tuple[float, float], ...]
    height_m: float
    height_estimated: bool = False
    osm_id: str | None = None
    tags: dict[str, str] = field(default_factory=dict)

    def bbox(self) -> tuple[float, float, float, float]:
        """(min_lat, min_lon, max_lat, max_lon)."""
        lats = [p[0] for p in self.ring]
        lons = [p[1] for p in self.ring]
        return (min(lats), min(lons), max(lats), max(lons))


def estimate_height_m(tags: dict[str, Any]) -> tuple[float, bool]:
    """OSM 태그 → (높이 m, 추정여부).

    우선순위: height > building:levels×층고 > 기본값.
    """
    raw_height = tags.get("height")
    if raw_height is not None:
        parsed = _parse_meters(raw_height)
        if parsed is not None:
            return (parsed, False)

    levels = tags.get("building:levels") or tags.get("levels")
    if levels is not None:
        try:
            n = float(str(levels).strip())
            if n > 0:
                return (n * DEFAULT_LEVEL_HEIGHT_M, True)
        except ValueError:
            pass

    return (DEFAULT_BUILDING_HEIGHT_M, True)


def _parse_meters(value: Any) -> float | None:
    """'12', '12 m', '12.5m' 등 → float(미터). 파싱 실패 시 None."""
    s = str(value).strip().lower().replace("meters", "").replace("metres", "").replace("m", "")
    s = s.strip()
    try:
        v = float(s)
        return v if v > 0 else None
    except ValueError:
        return None


def load_geojson(path: str) -> list[Building]:
    """GeoJSON(FeatureCollection) → 건물 리스트.

    Polygon/MultiPolygon Feature 의 외곽 링만 사용. 좌표는 [lon, lat] (GeoJSON 표준).
    properties 의 height/building:levels 로 높이를 추정한다.
    """
    with open(path, encoding="utf-8") as fh:
        data = json.load(fh)
    return _features_to_buildings(data.get("features", []))


def _features_to_buildings(features: list[dict]) -> list[Building]:
    buildings: list[Building] = []
    for feat in features:
        geom = feat.get("geometry") or {}
        props = feat.get("properties") or {}
        gtype = geom.get("type")
        coords = geom.get("coordinates") or []

        rings: list[list] = []
        if gtype == "Polygon" and coords:
            rings.append(coords[0])
        elif gtype == "MultiPolygon":
            rings.extend(poly[0] for poly in coords if poly)
        else:
            continue

        height, estimated = estimate_height_m(props)
        osm_id = props.get("id") or feat.get("id")
        tags = {k: str(v) for k, v in props.items()}
        for ring_coords in rings:
            ring = tuple((float(pt[1]), float(pt[0])) for pt in ring_coords if len(pt) >= 2)
            if len(ring) >= 3:
                buildings.append(
                    Building(
                        ring=ring,
                        height_m=height,
                        height_estimated=estimated,
                        osm_id=str(osm_id) if osm_id is not None else None,
                        tags=tags,
                    )
                )
    return buildings
